// ngx-lottie/esm2022/server/lottie-server.module.mjs
import { NgModule, InjectionToken, APP_INITIALIZER, TransferState, makeEnvironmentProviders, } from '@angular/core';
import { appInitializerFactory } from './internals';
import * as i0 from "@angular/core";
export const LOTTIE_SERVER_OPTIONS = new InjectionToken('LottieServerOptions');
export class LottieServerModule {
    static forRoot(options) {
        return {
            ngModule: LottieServerModule,
            providers: [
                {
                    provide: LOTTIE_SERVER_OPTIONS,
                    useValue: options,
                },
                {
                    provide: APP_INITIALIZER,
                    useFactory: appInitializerFactory,
                    multi: true,
                    deps: [LOTTIE_SERVER_OPTIONS, TransferState],
                },
            ],
        };
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LottieServerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.2.4", ngImport: i0, type: LottieServerModule }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LottieServerModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: LottieServerModule, decorators: [{
            type: NgModule
        }] });
export function provideLottieServerOptions(options) {
    return makeEnvironmentProviders([
        {
            provide: LOTTIE_SERVER_OPTIONS,
            useValue: options,
        },
        {
            provide: APP_INITIALIZER,
            useFactory: appInitializerFactory,
            multi: true,
            deps: [LOTTIE_SERVER_OPTIONS, TransferState],
        },
    ]);
}
